src/librustc/traits/select.rs

```diff
@@ -693,8 +705,9 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 match self.infcx
                     .subtype_predicate(&obligation.cause, obligation.param_env, p)
                 {
-                    Some(Ok(InferOk { obligations, .. })) => {
-                        self.evaluate_predicates_recursively(previous_stack, &obligations)
+                    Some(Ok(InferOk { mut obligations, .. })) => {
+                        self.add_depth(obligations.iter_mut(), obligation.recursion_depth);
```

Comment:
but it makes sense to increase the recursion depth *here* -- really, `subtype_predicate` should probably take the recursion depth as input, but I guess this is fine for now.